src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -119,78 +120,104 @@ SymbolKind.NetModule or
-                            }
-
-                            if (!merged)
-                            {
-                                // Just add it to the end
-                                intermediateProjects = intermediateProjects.Add((oldProjectState, translate));
-                            }
-                        }
-
-                        var newState = CompilationTrackerState.Create(
-                            withCompilationState.CompilationWithoutGeneratedDocuments,
-                            state.GeneratorInfo,
-                            finalCompilationWithGeneratedDocuments,
-                            intermediateProjects);
-                        return newState;
+                        return ForkWithCompilationTrackerState(oldProjectState, withCompilationTrackerState, translate);
                     }
-                    else if (state is NoCompilationState)
+                    else if (state is NoCompilationState noCompilationState)
                     {
-                        // We may still have a cached generator; we'll have to remember to run generators again since we are making some
-                        // change here. We'll also need to update the other state of the driver if appropriate.
-
-                        // The no compilation state can never be in the 'DocumentsAreFinal' state.  The only place where
-                        // we start with the NoCompilationState is the 'Empty' instance (where DocumentsAreFinal=false).
-                        // And then this is the only place where we get a NoCompilationState and create a new instance.
-                        // So there is no way to ever transition this to the DocumentsAreFinal=true state.
-                        Contract.ThrowIfTrue(state.GeneratorInfo.DocumentsAreFinal);
-                        var generatorInfo = state.GeneratorInfo;
-                        if (generatorInfo.Driver != null && translate != null)
-                            generatorInfo = generatorInfo.WithDriver(translate.TransformGeneratorDriver(generatorInfo.Driver));
-
-                        return new NoCompilationState(generatorInfo);
+                        return ForkNoCompilationState(noCompilationState, translate);
                     }
                     else
                     {
                         throw ExceptionUtilities.UnexpectedValue(state.GetType());
                     }
                 }
+
+                static NonFinalWithCompilationTrackerState ForkWithCompilationTrackerState(
+                    ProjectState oldProjectState,
+                    WithCompilationTrackerState state,
+                    CompilationAndGeneratorDriverTranslationAction? translate)
+                {
+                    var finalCompilationWithGeneratedDocuments = state is FinalCompilationTrackerState finalState
+                        ? finalState.FinalCompilationWithGeneratedDocuments
+                        : null;
```

Comment:
@jasonmalinowski i think this is technically not great.  ideally, even in the non-final sate, we'd preserve whatever other existing *stale compilation with generated documents* we might have.  afaict, that stale-compilation just acts as a hint, and as a way to bypass unnecessary computation when possible.  so it's always something we want to pass along.  but i'm not 100% sure on that and what to run that by you.

Reply:
Yeah this seems like we're going to lose that optimization here, generally -- if we transition to a final state and then drop it here on a fork, all the processing through the in-progress states is basically moot now.

But it seems the fix is straightforward, right? Something like (pardon the formatting, I'm typing this in GitHub)

```
var staleCompilationWithGeneratedDocuments =
    state is FinalCompilationTrackerState finalState ? finalState.FinalCompilationWithGeneratedDocuments :
    state is NonFinalWithCompilationTrackerState nonFinalState ? nonFinalState.StaleCompilationWithGeneratedDocuments : throw Unexpected("We covered all the types!");
```

Which your other renames I think make this clearer: the parameter on Create() is the stale compilation with documents, and in the final case the "stale" is just not very stale yet. :smile:

It almost may argue that StaleCompilationWithGeneratedDocuments can be pushed down further to WithCompilationTrackerState, and in the Final state case you just have a property that sounds not-stale that's really always the same instance.

Reply:
I think this probably got regressed with one of the earlier recent changes, since previously there might have been the very base "CompilationWithGeneratedDocuments" at a very base state that got pushed down to other types, but we lost the subtlety of it.